CIScripts/Common/Invoke-UntilSucceeds.ps1

```diff
@@ -29,34 +36,60 @@ function Invoke-UntilSucceeds {
         [Parameter(Mandatory=$true,
                    ValueFromPipeline=$true)] [ScriptBlock] $ScriptBlock,
         [Parameter(Mandatory=$false)] [int] $Interval = 1,
-        [Parameter(Mandatory=$true)] [int] $Duration = 3,
+        [Parameter(Mandatory=$false)] [int] $Duration,
+        [Parameter(Mandatory=$false)] [int] $NumRetries,
         [Parameter(Mandatory=$false)] [ScriptBlock] $Precondition,
         [Parameter(Mandatory=$false)] [String] $Name = "Invoke-UntilSucceds",
         [Switch] $AssumeTrue
     )
-    if ($Duration -lt $Interval) {
-        throw "Duration must be longer than interval"
+
+    if ((-not $Duration) -and (-not $NumRetries)) {
+        throw "Either non-zero -Duration or -NumRetries has to be specified"
+    }
+
+    if ($Duration) {
+        if ($NumRetries) {
+            throw "-Duration can't be used with -Retries"
+        }
+
+        if ($Duration -lt $Interval) {
+            throw "Duration must be longer than interval"
+        }
     }
+
     if ($Interval -eq 0) {
         throw "Interval must not be equal to zero"
     }
     $StartTime = Get-Date
 
+    $NumRetry = 0
+
     while ($true) {
-        $LastCheck = ((Get-Date) - $StartTime).TotalSeconds -ge $Duration
+        $NumRetry += 1
 
-        if ($Precondition -and -not (Invoke-Command $Precondition)) {
-            throw New-Object -TypeName CITimeoutException("Precondition was false, waiting aborted early")
+        $LastCheck = if ($Duration) {
+            ((Get-Date) - $StartTime).TotalSeconds -ge $Duration
+        } else {
+            $NumRetry -eq $NumRetries
         }
 
         try {
             $ReturnVal = Invoke-Command $ScriptBlock
+
+            if ($ReturnVal -as [StopRetrying]) {
+                throw [StopRetrying]::new()
+            }
+
             if ($AssumeTrue -Or $ReturnVal) {
                 return $ReturnVal
             } else {
-                throw New-Object -TypeName CITimeoutException("Did not evaluate to True." + 
+                throw New-Object -TypeName CITimeoutException("${Name}: Did not evaluate to True." + 
```

Comment:
Logic in this function got very complicated. Too many nested ifs and trycatches. Maybe refactor to a couple functions?